Added unimplemented expression cases in the ASTPrinter

Rajveer100 opened the pull request:
Part of #61601 

Filled out the remaining expression cases for ASTPrinter, which were left out after in the PR #40691.

Rajveer100 committed:
Added the expression cases of the form 'visit...Expr()'

Part of Issue #61601

Rajveer100 commented:
@hamishknight I have implemented one of the functions for you to review if I have done it the right way.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4785,7 +4787,13 @@ void PrintAST::visitErrorExpr(ErrorExpr *expr) {
 }
 
-void PrintAST::visitTernaryExpr(TernaryExpr *expr) {}
+void PrintAST::visitTernaryExpr(TernaryExpr *expr) {
+  if (expr->getCondExpr()) {
+    visit(expr->getThenExpr());
+  } else {
+    visit(expr->getElseExpr());
+  }
+}
```
This doesn't seem correct to me; to clarify, it should be outputting the original source code for the expression node. So in this case you'll want to first print the `getCondExpr`, then print a `?`, then print the `getThenExpr`, then print a `:`, then print the `getElseExpr`.

Note though that there is some awkwardness to the representation of an unresolved TernaryExpr (prior to operator folding), the `getElseExpr` may be null. In that case, we should just be able to skip printing it, and it should be handled by `SequenceExpr` printing.

Rajveer100 replied:
> Note though that there is some awkwardness to the representation of an unresolved TernaryExpr (prior to operator folding), the `getCondExpr` and `getElseExpr` may be null. In that case, we should just be able to skip printing them, and it should be handled by `SequenceExpr` printing.

Could you elaborate on `SequenceExpr`?

hamishknight replied:
> Could you elaborate on `SequenceExpr`?

SequenceExpr is a node we produce when parsing, and is used to represent a binary operator chain such as `x + y + z` (producing `{Expr, Op, Expr, Op, Expr}`), but also includes ternaries, where we produce a sequence of `{CondExpr, TernaryExpr, ElseExpr}`. Before type-checking, we fold SequenceExpr into the correct structure (e.g creating `BinaryExpr` nodes), for ternaries this involves setting the CondExpr and ElseExpr on the ternary rather than having them be elements of the sequence. This is done outside the parser as looking up operator precedences is required to correctly fold, and that requires doing name lookups.

So the correct way of printing this would be to check to make sure `getCondExpr` and `getElseExpr` are non-null when printing them. And then for SequenceExpr, we'll just want to print each element in turn.

I should note that SequenceExpr only exists in un-type-checked AST, so it won't appear with `-print-ast`, unless there is potentially a semantic error that prevents it from being folded. I think we still ought to handle it for completeness sake though.

Rajveer100 replied:
Thanks for the explanation!

hamishknight reviewed (changes requested):
Thanks for taking this on!

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4785,12 +4787,24 @@ void PrintAST::visitErrorExpr(ErrorExpr *expr) {
 
-void PrintAST::visitTernaryExpr(TernaryExpr *expr) {}
+void PrintAST::visitTernaryExpr(TernaryExpr *expr) {
+  visit(expr->getCondExpr());
+  Printer << " ? ";
+  visit(expr->getThenExpr());
+  Printer << " : ";
+  visit(expr->getElseExpr());
```
As mentioned, you'll want to check `getCondExpr` and `getElseExpr` to make sure they're non-null before printing

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4785,12 +4787,24 @@ void PrintAST::visitErrorExpr(ErrorExpr *expr) {
+}
 
 void PrintAST::visitIsExpr(IsExpr *expr) {
+  if (expr->getType()) {
+    Printer << "true";
+  } else {
+    Printer << "false";
+  }
```
I think there may be some misunderstanding here, to clarify; we're not trying to do any evaluation here, i.e we're not looking to produce a boolean result for an `IsExpr`. We are just printing the code, so this should look like `<expr> is <Type>`.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4785,12 +4787,26 @@ void PrintAST::visitErrorExpr(ErrorExpr *expr) {
 
-void PrintAST::visitTernaryExpr(TernaryExpr *expr) {}
+void PrintAST::visitTernaryExpr(TernaryExpr *expr) {
+  if (auto condExpr = expr->getCondExpr(); condExpr != nullptr) {
+    if (auto elseExpr = expr->getElseExpr(); elseExpr != nullptr) {
+      visit(expr->getCondExpr());
+      Printer << " ? ";
+      visit(expr->getThenExpr());
+      Printer << " : ";
+      visit(expr->getElseExpr());
+    }
+  }
```
Sorry I should have been clearer, we still want to print the `? ThenExpr :` even if the CondExpr and ElseExpr are null (as in that case they'll be printed by the SequenceExpr). Also `if (auto condExpr = expr->getCondExpr(); condExpr != nullptr)` is redundant, you just need `if (auto condExpr = expr->getCondExpr())` to check for null.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4785,12 +4787,26 @@ void PrintAST::visitErrorExpr(ErrorExpr *expr) {
+}
 
 void PrintAST::visitIsExpr(IsExpr *expr) {
+  visit(expr->getSubExpr());
+  Printer << " is ";
+  printType(expr->getCastType());
```
Looks good!

Rajveer100 commented:
How would `visitTypeJoinExpr` be handled?

hamishknight commented:
> How would `visitTypeJoinExpr`, `visitClosureExpr`, ... be handled?

Sorry, I shouldn't have included `TypeJoinExpr` and `OneWayExpr` in the list, they only exist during constraint solving, so I don't think we need any printing logic for them (`TypeJoinExpr` isn't even really representable as source code anyway).

`ClosureExpr` should print a closure e.g `{ <body> }` if there are no params, `{ <param>, ... in <body> }` if there are params, `{ (<param>, ...) -> <Result> in <body>}` if there is an explicit result type, `{ (<param>: <Type>, ...) in }` if there is an explicit param type. Note you'll also need to handle `CaptureListExpr`, which also prints a closure, but with a capture list e.g `{ [<capture>, ...] <param>, ... in }`.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4785,12 +4787,27 @@ void PrintAST::visitErrorExpr(ErrorExpr *expr) {
 }
 
 void PrintAST::visitTapExpr(TapExpr *expr) {
+  printType(expr->getVar()->getType());
+  printBraceStmt(expr->getBody());
```
`TapExpr` is a tricky beast because it contains a bunch of synthesized code that we probably don't want to print (I think ideally we'd just print an interpolated string literal as it appears in source, not as a chain of `append...` calls). There may be value in showing the `append...` calls, but I'm not sure it should be the default behavior. Additionally, without `do` expressions, we'd also have to resort to printing it as an immediately-executed closure.

To properly handle this, I think we'd need to jump through all the synthesized code to find the argument nodes for the `append...` calls, and use those to reconstruct the string literal segments and interpolated expressions. I'm okay if you want to leave this unimplemented for now.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4871,6 +4888,9 @@ void PrintAST::visitDictionaryExpr(DictionaryExpr *expr) {
 }
 
 void PrintAST::visitArrowExpr(ArrowExpr *expr) {
+  visit(expr->getArgsExpr());
+  Printer << " -> ";
+  visit(expr->getResultExpr());
```
For completeness, this should also print `async` and/or `throw` if `getAsyncLoc()` and/or `getThrowsLoc()` are valid.

Rajveer100 commented:
> > How would `visitTypeJoinExpr`, `visitClosureExpr`, ... be handled?
> 
> 
> 
> Sorry, I shouldn't have included `TypeJoinExpr` and `OneWayExpr` in the list, they only exist during constraint solving, so I don't think we need any printing logic for them (`TypeJoinExpr` isn't even really representable as source code anyway).
> 
> 
> 
> `ClosureExpr` should print a closure e.g `{ <body> }` if there are no params, `{ <param>, ... in <body> }` if there are params, `{ (<param>, ...) -> <Result> in <body>}` if there is an explicit result type, `{ (<param>: <Type>, ...) in }` if there is an explicit param type. Note you'll also need to handle `CaptureListExpr`, which also prints a closure, but with a capture list e.g `{ [<capture>, ...] <param>, ... in }`.

Alright! If I correctly understand, the capture in `closures` is similar to `lambda capture` in C++, just that in Swift it's handled by `weak self`? 

Also after the recent update in Swift, I presume we no longer have to use `self` keyword inside the body (to reference a capture variable) once it's `guarded`.

hamishknight commented:
> Alright! If I correctly understand, the capture in closures is similar to lambda capture in C++, just that in Swift it's handled by weak self?

Yes they are similar to lambda captures (specifically captures by value). `weak` and `unowned` can be used as ownership modifiers in a capture list, yes, but they can be applied to any capture, e.g:

```swift
class C {}
var c = C()
let fn = { [weak c] in }
```

And without them, the capture is by a strong reference:

```swift
class C {}
var c = C()
let fn = { [c] in }
```

Rajveer100 commented:
@hamishknight I have made some changes, please let me know if it's the right approach.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5028,12 +5053,35 @@ void PrintAST::visitBinaryExpr(BinaryExpr *expr) {
+    if (expr->hasExplicitResultType()) {
+      Printer << "(";
+    }
+    bool isFirst = true;
+    for (auto& parameter: *parameters) {
+      if (!isFirst) {
+        Printer << ", ";
+      }
+      visit(parameter);
+      isFirst = true;
```
LLVM actually has an `interleave` utility that makes it easier to interleave things between elements that you may want to use instead. Actually it seems they also have an `interleaveComma` utility specifically for cases like this.

Rajveer100 replied:
As of now, I am using the same printing logic, although, once we are down with all cases we can make changes as a touch up using the `llvm` utility you mentioned.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5050,6 +5098,13 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
 
 void PrintAST::visitKeyPathExpr(KeyPathExpr *expr) {
+  auto components = expr->getComponents();
+  for (int i = 0; i <= (int) components.size() - 1; i++) {
```
You can do `for (auto i : indices(components))` to iterate over the indices of a collection.

Rajveer100 replied:
Yeah, much better.

Rajveer100 replied:
But I used the normal iterative version to ensure the `'.'` in the `KeyPathExpr`.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5050,6 +5098,13 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
 }
 
 void PrintAST::visitKeyPathExpr(KeyPathExpr *expr) {
```
`KeyPathExpr` is also used to represent Obj-C key paths, e.g `#keyPath(...)`, you'll want to check `expr->isObjC()` for that

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5050,6 +5098,13 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
 void PrintAST::visitKeyPathExpr(KeyPathExpr *expr) {
+  auto components = expr->getComponents();
+  for (int i = 0; i <= (int) components.size() - 1; i++) {
+    printType(components[i].getComponentType());
```
The component type isn't how the component is spelled in source, you'll need to switch over the different component kinds and implement the printing logic for each of the kinds.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5050,6 +5098,13 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
 }
 
 void PrintAST::visitKeyPathExpr(KeyPathExpr *expr) {
+  auto components = expr->getComponents();
```
You will need to print the root type (`getRootType()`) before the components, e.g `\<RootType>.<component1>.<component2>`

hamishknight replied:
Not as important, but it would also be nice if this could handle printing key paths before the components are resolved. In that case, the components will be empty, and you'll want to print the `getParsedRoot()` and `getParsedPath()` instead.

Rajveer100 replied:
You mean to say that if `components` is `empty`, just after the `isObjC()`, we need to print `ParsedRoot` and `ParsedPath`?

hamishknight replied:
Yes

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4965,6 +4982,15 @@ void PrintAST::visitDictionaryExpr(DictionaryExpr *expr) {
 
 void PrintAST::visitArrowExpr(ArrowExpr *expr) {
+  visit(expr->getArgsExpr());
+  if (auto isAsync = expr->getAsyncLoc()) {
+    Printer << " async";
+  }
+  if (auto doesThrow = expr->getThrowsLoc()) {
+    Printer << " throws";
+  }
```
This works, but I think it would be better spelt as `if (expr->getAsyncLoc().isValid())` & `if (expr->getThrowsLoc().isValid())`.

hamishknight reviewed (commented):
Thanks for working more on this! I've left some more comments. It would great if you could start adding some tests to cover the printing logic you've added so far, take a look at the tests added in https://github.com/apple/swift/pull/40691/files for an example.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5028,12 +5053,35 @@ void PrintAST::visitBinaryExpr(BinaryExpr *expr) {
+    }
+  }
+  auto body = expr->getBody();
+  printBraceStmt(body);
```
Changed that.

Rajveer100 commented:
Please look into this [thread](https://github.com/apple/swift/commit/9f8f1d461c9e23d2314af10e1832bad77527fe07#r109779995)

Rajveer100 commented:
Few queries:

- For `visitSubscriptExpr()`, how do we print the `member` via `getMember()`? 
(Ex. In the expression `a[i]`, I am able to print the subscript itself, but not `'a'`.)

- Hint for starting `visitSequenceExpr()`.
- In `visitKeyPathExpr()`, what's the next step in the `switch case` part (in commit).

hamishknight commented:
Sorry for not getting back to this sooner, I should be able to take a look tomorrow. Note the feedback regarding adding test cases remains :)

Rajveer100 commented:
Hey that's fine! 
Thanks for your reply, please checkout this [issue](https://github.com/apple/swift/issues/64279#issuecomment-1525528546).

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5065,6 +5115,66 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+    Printer << "#keyPath(";
+  }
+  
+  if (auto root = expr->getRootType()) {
```
Methods like this will crash if `isObjC`, writing tests will help catch these issues :)

Rajveer100 replied:
You mean to say I need to `getRootType()` only if it is `!isObjC()`?

hamishknight replied:
Yes, for `isObjC` you only need to print the components

hamishknight commented:
> Hint for starting `visitSequenceExpr()`.

Should be able to print each of the elements in turn, with a space between them.

> In `visitKeyPathExpr()`, what's the next step in the switch case part (in commit).

You shouldn't need to do anything for `Identity`, `CodeCompletion`, and `OptionalWrap`.

- `(Unresolved)Property` is for a property component, e.g `\X.y`
- `(Unresolved)Subscript` is for a subscript e.g `\X.y[5]`
- `OptionalForce` is for a force unwrap e.g `\X.y!`
- `OptionalChain` is for an optional chain, e.g `\X.y?.z`
- `Identity` is for e.g `\X.self`
- `TupleElement` is for e.g `\(Int, Int).0`
- `DictionaryKey` has a different semantic meaning, but it has the same syntax as a property component, so that's how it should print

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5028,12 +5053,35 @@ void PrintAST::visitBinaryExpr(BinaryExpr *expr) {
 void PrintAST::visitClosureExpr(ClosureExpr *expr) {
+  Printer << "{ ";
+  if (auto parameters = expr->getParameters()) {
+    if (expr->hasExplicitResultType()) {
+      Printer << "(";
+    }
```
I have to fix a few things in `visitClosureExpr()`, will do it, while I figure out and add the tests.

hamishknight commented:
You should be able to use `printType` for that

Rajveer100 commented:
That didn't work, I actually got it to work using `rootType->print(Printer, PrintOptions())`.

hamishknight commented:
What error did you get with `printType`? It's better to use `printType`, because that will use a consistent set of options that is shared by the rest of the ASTPrinter

Rajveer100 commented:
> The `visit()` function throws unexpected errors when passing the `rootType` (i.e. `expr->getRootType()`) , although it does allow a `TypeRepr`. 
> 
> 
> 
> `error: no member named 'visitTypeRepr' in '(anonymous namespace)::PrintAST'`

Here.

Rajveer100 commented:
`No viable conversion from 'swift::TypeRepr *' to 'swift::Type'`

hamishknight commented:
Ah sorry, in that case you'll want to use `printTypeLoc`

Rajveer100 commented:
That works! 
By the way, I have started adding tests.

Rajveer100 commented:
Sure, I guess I will finish the main ones first, we may then discuss about the rest.

Rajveer100 commented:
I need a little description about the following expressions:
- visitEnumIsCaseExpr
- visitKeyPathDotExpr

There are few more which I would ping you about later!

hamishknight commented:
Sorry for the delay on this!

> `visitEnumIsCaseExpr`

This is synthesized in certain cases when type-checking `<expr> is <Type>`, we turn it into `(<expr> as? <Type>) != nil`. A faithful printing implementation could print it as that, but I think ideally we'd try and print it using the `is` syntax instead.

> `visitKeyPathDotExpr`

This is constructed by the parser to represent the initial `.` in a keypath. You should be able to just print it as `.`.

Rajveer100 commented:
Keeping the below list for reference to what is complete (or close):

 - [x] IsExpr
 - [x] ArrowExpr
 - [x] CoerceExpr
 - [x] TenaryExpr
 - [x] ClosureExpr
 - [ ] CaptureListExpr
 - [x] KeyPathExpr
 - [ ] KeyPathApplicationExpr
 - [ ] SubscriptExpr
 - [ ] BindOptionalExpr
 - [ ] ForceValueExpr
 - [ ] OptionalTryExpr
 - [ ] PrefixUnaryExpr
 - [ ] PostfixUnaryExpr
 - [ ] TupleElementExpr
 - [ ] InterpolatedStringLiteralExpr
 - [ ] TapExpr
 - [ ] SuperRefExpr
 - [ ] DynamicTypeExpr
 - [ ] ObjCSelectorExpr
 - [ ] ObjectLiteralExpr
 - [ ] MagicIdentifierLiteralExpr
 - [ ] DynamicMemberRefExpr
 - [ ] DynamicSubscriptExpr
 - [ ] EnumIsCaseExpr
 - [ ] KeyPathDotExpr
 - [ ] UnresolvedMemberExpr
 - [ ] EditorPlaceholderExpr
 - [ ] OverloadedDeclRefExpr
 - [ ] SequenceExpr
 - [ ] UnresolvedDeclRefExpr
 - [ ] UnresolvedPatternExpr
 - [ ] DotSyntaxBaseIgnoredExpr
 - [ ] OtherConstructorDeclRefExpr
 - [ ] RebindSelfInConstructorExpr
 - [ ] MakeTemporarilyEscapableExpr
 - [ ] All the ImplicitConversionExpr subclasses, we should stamp out their definitions as printing their sub-expression using a macro

Rajveer100 commented:
@hamishknight 
I have made few more changes, regarding few tests, I am unsure of some of the kind of expressions, while I am progressively adding the ones I am clear with.

hamishknight commented:
@Rajveer100 Would you like me to do another review pass over what you have so far? Adding more tests for the additional AST nodes you've implemented printing logic for is definitely a good idea, as it will expose any issues with the implementation, which you can iterate on. Let me know if you have any questions regarding what syntax particular AST nodes have.

Rajveer100 commented:
@hamishknight 
I am making few more changes for the respective methods before you can do another pass of review!

I will reach to you in case I have any specific queries for the AST expressions!

Rajveer100 commented:
Could you briefly describe these expressions:

- `RebindSelfInConstructorExpr`
- `MakeTemporarilyEscapableExpr`

Once I implement them, I will query the next expressions. (probably in 2s)

hamishknight commented:
@Rajveer100 Sorry again for the late reply! Printing the `expr->getFn()` is correct, to fix the printing of `Int.-x`, you'll want to change the printing logic of `DotSyntaxCallExpr`, we should be checking to see if `expr->getFn()` is for an operator decl, and if so we shouldn't print the base. You can call `->getReferencedDecl()` to get the referenced decl from the expression, and then check for `->isOperator()` on the returned decl.

Rajveer100 commented:
Cool, Thanks!

Rajveer100 commented:
Below statement throws an error (BAD_ACCESS), specifically when calling `getDecl()`:

`auto decl = expr->getReferencedDecl().getDecl();`

Used above to get the `isOperator()`.

hamishknight commented:
@Rajveer100 I don't think calling `getDecl()` should cause a crash, but calling `isOperator()` on a null Decl would certainly cause a crash, you'll want to check for `nullptr` before checking `isOperator()`, as there may not be a directly referenced decl available (in which case it definitely isn't for an operator).

Rajveer100 commented:
Well, `getDecl()` is in fact causing the crash, I re-confirmed it to be sure!

hamishknight commented:
Hmm, does the crash still happen if you change the code to the following?

```cxx
  auto decl = expr->getFn()->getReferencedDecl().getDecl();
  if (decl && !decl->isOperator())
```

Note also you need to check `expr->getFn()` for the `DotSyntaxCallExpr`

Rajveer100 commented:
Hey that works, although I was wondering, there already exists a `visitXXX` for `DotSyntaxBaseIgnoredExpr`, so why doesn't the compiler implicitly use that instead of calling the `DotSyntaxExpr`.

Could we possibly improve this?

hamishknight commented:
> Hey that works, although I was wondering, there already exists a `visitXXX` for `DotSyntaxBaseIgnoredExpr`, so why doesn't the compiler implicitly use that (if it knows it's an operator) instead of calling the `DotSyntaxExpr`.

Due to the way we currently model static methods, SILGen expects us to provide a metatype value for `self`, so `DotSyntaxBaseIgnoredExpr` wouldn't be correct as we'd be throwing away `self`. We could introduce a new node specifically for operators, but that's probably not worth it.

Rajveer100 commented:
For the `InterpolatedStringLiteralExpr`, how can we retrieve the respective `tokens` (i.e. `\(..)`), and the `regular literals` itself within the string? Are there methods that specifically differentiate between them while traversing?

Rajveer100 commented:
@hamishknight 
I think it's time for a review, you may refer the [checklist](https://github.com/apple/swift/pull/64667#issuecomment-1602197263) for an overview. It would be great if you could briefly describe the unchecked items. I have also left few `blank` lines where I was unsure of some `AST Expr`. While I am continually adding `tests` for validation I might have missed out a few!

hamishknight commented:
Let me maintain my own list of nodes I have approved the implementation of :) Just to help subsequent reviews so I don't have to look at everything again.

 - [ ] IsExpr
 - [ ] ArrowExpr
 - [ ] CoerceExpr
 - [ ] TenaryExpr
 - [ ] ClosureExpr
 - [ ] CaptureListExpr
 - [ ] KeyPathExpr
 - [ ] KeyPathApplicationExpr
 - [ ] SubscriptExpr
 - [ ] BindOptionalExpr
 - [ ] ForceValueExpr
 - [ ] OptionalTryExpr
 - [ ] PrefixUnaryExpr
 - [ ] PostfixUnaryExpr
 - [ ] TupleElementExpr
 - [ ] InterpolatedStringLiteralExpr
 - [ ] TapExpr
 - [ ] SuperRefExpr
 - [ ] DynamicTypeExpr
 - [ ] ObjCSelectorExpr
 - [ ] ObjectLiteralExpr
 - [ ] MagicIdentifierLiteralExpr
 - [ ] DynamicMemberRefExpr
 - [ ] DynamicSubscriptExpr
 - [ ] EnumIsCaseExpr
 - [ ] KeyPathDotExpr
 - [ ] UnresolvedMemberExpr
 - [ ] EditorPlaceholderExpr
 - [ ] OverloadedDeclRefExpr
 - [ ] SequenceExpr
 - [ ] UnresolvedDeclRefExpr
 - [ ] UnresolvedPatternExpr
 - [ ] DotSyntaxBaseIgnoredExpr
 - [ ] OtherConstructorDeclRefExpr
 - [ ] RebindSelfInConstructorExpr
 - [ ] MakeTemporarilyEscapableExpr
 - [ ] All the ImplicitConversionExpr subclasses

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5408,9 +5604,12 @@ void PrintAST::visitOtherConstructorDeclRefExpr(OtherConstructorDeclRefExpr *exp
 }
 
 void PrintAST::visitMakeTemporarilyEscapableExpr(MakeTemporarilyEscapableExpr *expr) {
+  visit(expr->getSubExpr());
```
You'll need to reconstruct the original call here, e.g:

```swift
func foo(_ fn: () -> Void) {
  withoutActuallyEscaping(fn) { fn in
    // ...
  }
}
```

hamishknight replied:
Oh looks like we actually keep track of the original expr, you should be able to print `expr->getOriginalExpr()` here.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5426,6 +5625,7 @@ void PrintAST::visitCovariantReturnConversionExpr(CovariantReturnConversionExpr
 }
 
 void PrintAST::visitInterpolatedStringLiteralExpr(InterpolatedStringLiteralExpr *expr) {
+  visit(expr->getInterpolationExpr());
```
As mentioned previously, implementing this will require traversing the TapExpr to reconstruct the string interpolation syntax.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5112,8 +5137,7 @@ void PrintAST::visitTupleExpr(TupleExpr *expr) {
 
-void PrintAST::visitTypeJoinExpr(TypeJoinExpr *expr) {
-}
+void PrintAST::visitTypeJoinExpr(TypeJoinExpr *expr) { }
```
Might be worth adding an `llvm_unreachable` in the body of this, since we should never see a TypeJoinExpr outside of the constraint system.

Rajveer100 replied:
Cool.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5134,12 +5158,35 @@ void PrintAST::visitBinaryExpr(BinaryExpr *expr) {
 
-void PrintAST::visitOneWayExpr(OneWayExpr *expr) {
-}
+void PrintAST::visitOneWayExpr(OneWayExpr *expr) { }
```
Also might be worth adding an `llvm_unreachable` here since we should never see it outside the constraint system

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+  
+  auto components = expr->getComponents();
+  if (!components.empty()) {
+    for (auto& component: components) {
```
```suggestion
    for (auto &component: components) {
```

hamishknight replied:
Also I think it would be great if you could factor out the component printing logic, which should make this function a bit easier to read, as e.g for the Obj-C case you could do:

```cxx
  if (expr->isObjC()) {
    Printer << "#keyPath(";
    printKeyPathComponents(expr->getComponents());
    Printer << ")";
    return;
  }
```

then you don't have to worry about intermixing it with the regular KeyPath printing logic.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+          Printer << ".";
+          
+          
+          Printer << "[";
+          // component.forUnresolvedSubscript(<#ASTContext &ctx#>, args)
+          Printer << "]";
```
You'll want to print the arguments here, similar to printing for SubscriptExpr. In fact it may be worth factoring out an ArgumentList printing function at this point that you could share between CallExpr, SubscriptExpr, and here. You could have it take a `bool forSubscript` argument to change printing from `(...)` to `[...]`.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+        }
+        case swift::KeyPathExpr::Component::Kind::UnresolvedSubscript: {
+          auto args = component.getSubscriptArgs();
+          Printer << ".";
```
KeyPath subscripts _can_ appear with a dot for the first component, but otherwise they need to be without the dot. I think it's probably best to print without the dot in all cases though.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+        }
+        case swift::KeyPathExpr::Component::Kind::Property: {
+          Printer << ".";
+          component.getDeclRef().getDecl()->print(Printer, PrintOptions());
```
Printing the decl isn't correct here, you only want to print the name

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+        }
+        case swift::KeyPathExpr::Component::Kind::Subscript: {
+          auto args = component.getSubscriptArgs();
+          Printer << ".";
```
Similar comment about the dot

Rajveer100 replied:
Yep.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+          Printer << ".";
+          
+          
+          Printer << "[";
+          // component.forSubscript(<#ASTContext &ctx#>, component.getDeclRef(), args, component.getComponentType(), component.getSubscriptIndexHashableConformances());
+          Printer << "]";
```
And about the arguments :)

Rajveer100 replied:
Yeah.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+          break;
+        }
+        case swift::KeyPathExpr::Component::Kind::TupleElement: {
+          Printer << "(";
+          
+          Printer << ")";
```
These shouldn't be needed, this is just for the e.g `.0` part

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+          break;
+        }
+        case swift::KeyPathExpr::Component::Kind::Identity: {
+          break;
```
This is `.self`

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
 }
 
 void PrintAST::visitKeyPathExpr(KeyPathExpr *expr) {
```
Could you add some KeyPath tests please :)

hamishknight reviewed (commented):
Some good progress here!

hamishknight commented:
- KeyPathApplicationExpr
 	- This is an application of a key path that has the form `x[keyPath: kp]`
 - TupleElementExpr
 	- An element access on a tuple e.g `x.0`
 - InterpolatedStringLiteralExpr
 	- An interpolated string e.g `"literal \(expr) more literal \(anotherExpr)"` 
 - TapExpr
 	- This should be handled as a part of the interpolated string literal printing, it doesn't have a printed representation outside of that
 - ObjCSelectorExpr
 	- This is `#selector(...)`, I think you already have this one
 - ObjectLiteralExpr
 	- This is a magic literal such as `#colorLiteral(...)`, you should be able to print using `getLiteralKindRawName`, plus printing the argument list
 - MagicIdentifierLiteralExpr
 	- This is a magic literal such as `#file`, you should be able to print using `MagicIdentifierLiteralExpr::getKindString`
 - DynamicMemberRefExpr
 	- This is for a member access on an AnyObject base, it should be printed the same as a regular member lookup
 - DynamicSubscriptExpr
 	- This is for a subscript on an AnyObject base, it should be printed the same as a normal subscript
 - KeyPathDotExpr
 	- This represents a `.` in a key path before its components are resolved
 - UnresolvedMemberExpr
 	- This is for an unresolved dot lookup, e.g `.foo` 
 - EditorPlaceholderExpr
	- This is for an editor placeholder `<#placeholder#>`, they can also include a type, in which case they print as `<#T##placeholder##Type#>` where `placeholder` can be any text, and `Type` is a type.
 - OverloadedDeclRefExpr
 	- This is a reference to a set of overloaded decls, it should be printable as the base name of the first decl in the list, though we probably ought to also check for `getNameLoc()->isCompound()`, and print the full name with argument labels in that case. 
 - SequenceExpr
 	- This is for an operator chain prior to type-checking where it gets folded, as mentioned previously it should be printable as the elements separated by a space.
 - OtherConstructorDeclRefExpr
	- This is a reference to an initializer within another initializer, I think it should be printable as `init`. 
 - RebindSelfInConstructorExpr
	- This is synthesized and wraps calls to `self.init` in initializers, as previously mentioned, I don't think we need special printing logic for it, we should just have to print the underlying node.
 - MakeTemporarilyEscapableExpr
 	- This is created by the type-checker to represent a call to the standard library function, e.g:
    ```swift
    func foo(_ fn: () -> Void) {
        withoutActuallyEscaping(fn) { fn in
            // ...
        }
    }
	```
   
 - All the ImplicitConversionExpr subclasses
 	- As previously discussed, these should be stamped out with macros, and should just print the underlying node.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5134,12 +5158,35 @@ void PrintAST::visitBinaryExpr(BinaryExpr *expr) {
+void PrintAST::visitOneWayExpr(OneWayExpr *expr) { }
 
 void PrintAST::visitClosureExpr(ClosureExpr *expr) {
+  Printer << "{ ";
```
Today I learnt we also support attributes for closures, which we also ought to print here by printing its `getAttrs()`, this allows e.g `{ @MainActor in }`

Rajveer100 commented on test/expr/print/func_closures.swift:
```diff
@@ -0,0 +1,11 @@
+// CHECK: internal func fetch() async throws -> String {
+// CHECK: }
+
+let fn = { (x: Int) in
+}
```
Sounds good! I have added the `single expression` case as well.

Rajveer100 commented on test/expr/print/misc_expr.swift:
```diff
@@ -0,0 +1,36 @@
+  a[i] = i;
+}
+// CHECK: @_hasInitialValue internal var a: [Int] = [1, 3, 2]
+// CHECK: for i in  {
```
Yeah, I have to implement it.

Rajveer100 replied:
I have implemented the sequence expression, although it still doesn't reflect for the `[start]...[end]`, or any sort of sequences. 

Is there any other expression which is also responsible?

hamishknight replied:
Sorry I should have clarified what I meant by "sequence expression" here, I didn't mean SequenceExpr, I meant the sequence expression of ForEachStmt, which can be any Expr, not just SequenceExpr :) It should be printable as something like the following:

```cxx
  if (auto *seq = stmt->getTypeCheckedSequence()) {
    // Look through the call to '.makeIterator()'
    if (auto *CE = dyn_cast<CallExpr>(seq)) {
      if (auto *SAE = dyn_cast<SelfApplyExpr>(CE->getFn()))
        seq = SAE->getBase();
    }
    visit(seq);
  } else {
    visit(stmt->getParsedSequence());
  }
```

We need to strip off the `.makeIterator()` call in the type-checked case. Alternatively, we could store just the type-checked sequence expression on the node to make retrieving it easier for printing.

Rajveer100 commented on test/expr/print/misc_expr.swift:
```diff
@@ -0,0 +1,36 @@
+}
+// CHECK: @_inheritsConvenienceInitializers internal class B : A {
+// CHECK:   override internal init() {
+// CHECK:     self.()
```
Regarding this, does it only depend on the `SuperRefExpr` or am I missing implementation of some other `expr` as right now I am only printing the `"super"` keyword and it doesn't reflect here, instead printing `'self'`. Going through the codebase, it seems like the `super.x` part worked correctly as it's a `member ref` while the other is rather a `method` potentially?

Not sure how is the `'init'` part handled, is there any existing `expr` that checks whether it's an `initialiser`?

hamishknight replied:
The `init` part should be handled by printing OtherConstructorDeclRefExpr as `init` I think. I think the reason you were getting `self` was due to your prior implementation of RebindSelfInConstructorExpr, try running the test again with the new implementation :)

Rajveer100 replied:
The `init` keyword in `OtherConstructorDeclRefExpr` seems to work well, although I do not see the `super` keyword reflecting from the `RebindSelfInConstructorExpr`, which for now I have done a simple `visit` call for the sub expression.

At the moment, I get the following output:

```Swift
@_inheritsConvenienceInitializers internal class B : A {
  override internal init() {
    init()
    super.x = 2

  }
  @objc deinit {
  }
}
```

Rajveer100 replied:
Cool, that works.

Rajveer100 commented on test/expr/print/conditions.swift:
```diff
@@ -36,3 +36,24 @@ repeat {
+// CHECK: if x is Int {
+// CHECK: }
+
+enum someError: Error {
```
Oh yeah! I would say that's not a `nit`, rather critical :)! Fixed.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5134,12 +5158,35 @@ void PrintAST::visitBinaryExpr(BinaryExpr *expr) {
 }
 
 void PrintAST::visitCoerceExpr(CoerceExpr *expr) {
+  visit(expr->getSubExpr());
+  Printer << " as ";
+  printType(expr->getCastType());
```
Added.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5134,12 +5158,35 @@ void PrintAST::visitBinaryExpr(BinaryExpr *expr) {
+  if (auto parameters = expr->getParameters()) {
+    Printer << "(";
+    bool isFirst = true;
+    for (auto& parameter: *parameters) {
```
Does `git-clang-format` make a change in such cases, or do I have to do it manually?

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5168,9 +5312,19 @@ void PrintAST::visitForceTryExpr(ForceTryExpr *expr) {
+  int len = args->size() - 1;
+  for (auto arg: *args) {
+    printArgument(arg);
+    if (len-- != 0)
+      Printer << " ";
```
The `variable` logic won't work here, as we can't check for the last element (`isLast`), so I think I will have to go with `llvm::interleave`, and probably replicate the same for the rest of the `expressions` for code coverage and consistency.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5168,9 +5312,19 @@ void PrintAST::visitForceTryExpr(ForceTryExpr *expr) {
 }
 
 void PrintAST::visitSuperRefExpr(SuperRefExpr *expr) {
+  if (expr->getSuperLoc().isValid()) {
```
Let's keep the discussion related to `SuperRefExpr` in this [thread](https://github.com/apple/swift/pull/64667#discussion_r1292226651).

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5168,9 +5312,19 @@ void PrintAST::visitForceTryExpr(ForceTryExpr *expr) {
 }
 
 void PrintAST::visitSequenceExpr(SequenceExpr *expr) {
+  auto args = expr->getArgs();
```
Yep.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5180,46 +5334,82 @@ void PrintAST::visitMemberRefExpr(MemberRefExpr *expr) {
 }
 
 void PrintAST::visitEnumIsCaseExpr(EnumIsCaseExpr *expr) {
+  visit(expr);
```
Cool.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5381,6 +5576,7 @@ void PrintAST::visitUnevaluatedInstanceExpr(UnevaluatedInstanceExpr *expr) {
 }
 
 void PrintAST::visitDotSyntaxBaseIgnoredExpr(DotSyntaxBaseIgnoredExpr *expr) {
+  visit(expr);
```
Will add a test for it to check.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5229,7 +5419,10 @@ void PrintAST::visitDerivedToBaseExpr(DerivedToBaseExpr *expr) {
+  if (decl && !decl->isOperator()) {
+    visit(expr->getBase());
+  }
   Printer << ".";
```
I have corrected it for the appropriate behaviour.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5180,46 +5334,82 @@ void PrintAST::visitMemberRefExpr(MemberRefExpr *expr) {
 }
 
 void PrintAST::visitPrefixUnaryExpr(PrefixUnaryExpr *expr) {
+  visit(expr->getFn());
+  visit(expr->getOperand());
```
Added.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5180,46 +5334,82 @@ void PrintAST::visitMemberRefExpr(MemberRefExpr *expr) {
 }
 
 void PrintAST::visitBindOptionalExpr(BindOptionalExpr *expr) {
+  visit(expr->getSubExpr());
+  Printer << "?";
```
Added.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5180,46 +5334,82 @@ void PrintAST::visitMemberRefExpr(MemberRefExpr *expr) {
 }
 
 void PrintAST::visitPostfixUnaryExpr(PostfixUnaryExpr *expr) {
+  visit(expr->getOperand());
+  visit(expr->getFn());
```
Added.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5180,46 +5334,82 @@ void PrintAST::visitMemberRefExpr(MemberRefExpr *expr) {
 }
 
 void PrintAST::visitObjCSelectorExpr(ObjCSelectorExpr *expr) {
+  Printer << "#selector";
+  Printer << "(";
+  visit(expr->getSubExpr());
+  Printer << ")";
```
How do we use modules such as `Foundation` and other `SDKs` to use the `@objc` and `#selector` for this test. I tried adding the `-sdk [options] flag`, although I didn't seem to make it work.

hamishknight replied:
You should be able to use `-disable-objc-attr-requires-foundation-module` to avoid having to import Foundation

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5408,9 +5604,12 @@ void PrintAST::visitOtherConstructorDeclRefExpr(OtherConstructorDeclRefExpr *exp
 }
 
 void PrintAST::visitRebindSelfInConstructorExpr(RebindSelfInConstructorExpr *expr) {
+  Printer << expr->getSelf()->getName();
```
Sure.

hamishknight commented:
FYI I'm out on vacation this week, I'll be sure to reply to your comments next week :)

Rajveer100 commented:
Sure!

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5180,46 +5334,82 @@ void PrintAST::visitMemberRefExpr(MemberRefExpr *expr) {
+  Printer << "[";
+  auto isFirst = true;
+  for (auto& par: captureList) {
+    Printer << par.getVar()->getName();
```
How would one check the existence of `weak` and `bindings` for a `captured` variable ? 
Until now, I seemed to have found `isWeakImported(<#ModuleDecl *fromModule#>)`, but I think its for modules referring `@_weaklinked` attribute.

hamishknight replied:
In the CaptureListEntry will be the PatternBindingDecl for the binding, you'll want to print both the variable and the initializer (you could skip printing the initializer if you see that it's a DeclRefExpr with the same name as the variable, but it's not crucial). For modifiers such as `weak`, I think you should be able to use `printAttributes` on the given VarDecl.

Rajveer100 commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+    for (auto& component: components) {
+      auto kind = component.getKind();
+      switch (kind) {
+        case swift::KeyPathExpr::Component::Kind::Invalid: {
```
Sure, made the changes.

Rajveer100 commented:
Could you potentially provide a documentation/guide to the compiler architecture, so I can meanwhile grasp about the source code while I contribute?

This would help me in knowing each of the individual units and make me more familiar!

Rajveer100 commented:
@hamishknight @hamishknight 
I am currently refactoring the printing logic of `KeyPathExpr` and at the moment the `visitXXXExpr` definitions are under the `PrintAST` namespace, so how what would be the right way to create a function like `printKeyPathComponents`.

My idea is to add a new `#define` creating the definition like the `visit` methods using `macros,` what's your suggestion?

hamishknight commented:
> I am currently refactoring the printing logic of `KeyPathExpr` and at the moment the `visitXXXExpr` definitions are under the `PrintAST` namespace, so how what would be the right way to create a function like `printKeyPathComponents`, since I cannot use the `Printer` and many more such private members.

I think you should be able to define them as members of `PrintAST`, see e.g `printAccess`

hamishknight commented:
> Could you potentially provide a doc/guide to the compiler architecture, so I can meanwhile grasp about the source code while I contribute?

While mainly focussed on the generics system, https://download.swift.org/docs/assets/generics.pdf has a good overview of the overall compiler pipeline in chapter 2, which you may be interested in. There are also a bunch of useful links in https://github.com/apple/swift/blob/main/docs/ExternalResources.md for articles on various parts of the compiler.

Rajveer100 commented:
Thanks!

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5180,46 +5334,82 @@ void PrintAST::visitMemberRefExpr(MemberRefExpr *expr) {
 }
 
 void PrintAST::visitCaptureListExpr(CaptureListExpr *expr) {
+  auto captureList = expr->getCaptureList();
```
`AbstractClosureExpr` is the base class that's inherited by both `ClosureExpr` (for explicitly written closures in the source), and for `AutoClosureExpr` (for implicitly created closures). We need to be able to handle both because both can have capture lists, and so `captureList->getClosureBody()` returns an `AbstractClosureExpr`. IIRC the main case where we introduce capture lists for auto-closures is for key-paths as functions. We may want to try detect that case and print using the key-path syntax instead, but I think for now the closure syntax is fine.

Rajveer100 commented:
@hamishknight 
What's the right way to pass arguments for using `objc selectors`, I used the following:

```
/Debug.swift -print-ast -disable-objc-attr-requires-foundation-module
``` 

```
error: import the 'ObjectiveC' module to use '#selector'
let action = #selector(MyViewController.tappedButton)
             ^
```

Also, how do we use `#colorliteral`, since that also causes errors:
```
could not infer type of color literal
let myColor = #colorLiteral(red: 0.292, green: 0.081, blue: 0.6, alpha: 255)
              ^
```

Rajveer100 commented:
Also, here's the updated [checklist](https://github.com/apple/swift/pull/64667#issuecomment-1602197263) from my end, with just a few `expressions` left out and `tests` for some of them.

hamishknight commented:
> What's the right way to pass arguments for using objc selectors, I used the following:
> 
> ```
> /Debug.swift -print-ast -disable-objc-attr-requires-foundation-module
> ```
> 
> ```
> error: import the 'ObjectiveC' module to use '#selector'
> let action = #selector(MyViewController.tappedButton)
>              ^
> ```

Oh interesting looks like I was mistaken. In that case you'll want to split out selector printing into a separate test file, and there you can do `import ObjectiveC`, and add `// REQUIRES: objc_interop` to the test. That should be sufficient to get the test working.

> Also, how do we use #colorliteral, since that also causes errors:

You should be able to define a mock color type in the test, e.g:

```swift
struct FakeColor: _ExpressibleByColorLiteral {
  init(_colorLiteralRed: Float, green: Float, blue: Float, alpha: Float) {}
}
typealias _ColorLiteralType = FakeColor
```

That should solve the issue.

Rajveer100 commented:
> You should be able to define a mock color type in the test, e.g:
> 
> ```swift
> struct FakeColor: _ExpressibleByColorLiteral {
>   init(_colorLiteralRed: Float, green: Float, blue: Float, alpha: Float) {}
> }
> typealias _ColorLiteralType = FakeColor
> ```
> 
> That should solve the issue.

Is there an addition to this? 
This itself doesn't reflect any change by the `visit` method for the `expression` when commenting out the implementation.
I get the same result as is.

hamishknight commented:
Does it not resolve the `could not infer type of color literal` error? Or is the issue that the printing isn't working? Because that sounds like a different issue

Rajveer100 commented:
Yes, the previous issue is resolved, this the next issue where I am unable to test my implementation.

Rajveer100 commented:
No, that's not the issue, what I meant is, when I comment out my implementation, I see no difference in the output for the snippet you shared above when I uncomment the same.

hamishknight commented:
> No, that's not the issue, what I meant is, when I comment out my implementation, I see no difference in the output for the snippet you shared above when I uncomment the same.

I might be misunderstanding you, but the snippet I shared is only necessary to avoid the compiler error, it otherwise doesn't affect the printing of a `#colorLiteral`. Likewise, the printing logic of `#colorLiteral` won't affect the printed output for the `FakeColor` struct (and you don't have to test the printing of that since it's only needed to assist the testing of `#colorLiteral` printing).

hamishknight commented:
> What's the suitable test for this expression though? As directly using `#colorliteral` didn't work out and the snippet you shared though resolves the compile error doesn't allow me to test for the same

What issue are you running into? AFAIK you should be able to test the same way the other expressions are tested

Rajveer100 commented:
Apologies! 
I misunderstood the entire thing, ok so just to summarise what I interpreted from the snippet is that you are created your own custom` colour literal` that `conforms` to the `protocol`, and I would initialise a new variable this way:

```Swift
let color = FakeColor(_colorLiteralRed: 0.1, green: 1.0, blue: 1.1, alpha: 2.3)
```

When I tested the code, the `visit` function didn't even affect the output at all, because I just `replaced` my previous `snippet` with this, i.e, removed it which means I thought your snippet is an entire `replacement`, since we cannot use the following statement and the only possible way test is by using our own `custom initialiser`:

```Swift
let myColor = #colorLiteral(red: 0.292, green: 0.081, blue: 0.6, alpha: 255)
```

I then realised that it's in fact an addition, which allows us to use `#colorliteral` for declaration, which if I correctly understand, due to the creation of the`typealias`. 

:)

hamishknight commented:
Yes that's correct, sorry I probably should have explained a bit more of what the snippet did :)

Rajveer100 commented:
Can you have a look at the changes for `EditorPlaceholderExpr` which I have recently pushed, not sure if that's the intended implementation?

Also, these are few things that remain from my perspective:

- `TapExpr`
- `printClosure` (for `ClosureExpr` and `CaptureExpr` as a common printing logic)
- Sequence expression for `ForEachStmt` Expr
- Few remaining `tests`
- Handling `ImplicitConversionExpr` and other subclasses

Let me know if I am missing out here!

Rajveer100 commented:
@hamishknight

hamishknight commented:
Sorry again, I should have some time today to take a look :)

hamishknight commented on test/expr/print/misc_expr.swift:
```diff
@@ -0,0 +1,117 @@
+typealias _ColorLiteralType = FakeColor
+
+let myColor = #colorLiteral(red: 0.292, green: 0.081, blue: 0.6, alpha: 255)
+// CHECK: @_hasInitialValue internal let myColor: FakeColor = colorLiteral(red: 0.292, green: 0.081, blue: 0.6, alpha: 255)
```
I think it would be great if we could change these tests so they also do a typecheck of the output to ensure it's valid Swift, as that would help avoid these kinds of issues. You should be able to write it as something like:

```
// RUN: %empty-directory(%t)
// RUN: %target-swift-frontend -print-ast %s > %t/result.swift
// RUN: %target-typecheck-verify-swift %t/result.swift
// RUN: %FileCheck %s < %t/result.swift
```

Rajveer100 replied:
Do you mean to add this for every `test` file w.r.t to `ASTPrinter`?

hamishknight replied:
Ideally yes, there may be some constructs that it won't work for, e.g memberwise initializers since they have their bodies generated by SILGen, but we could extract that testing out to a separate file that doesn't do a typecheck of the result.

hamishknight commented:
Reposting the list of nodes I've approved so I don't have to scroll up for it

 - [x] IsExpr
 - [x] ArrowExpr
 - [x] CoerceExpr
 - [x] TenaryExpr
 - [ ] ClosureExpr
 - [ ] CaptureListExpr
 - [ ] KeyPathExpr
 - [ ] KeyPathApplicationExpr
 - [ ] SubscriptExpr
 - [x] BindOptionalExpr
 - [x] ForceValueExpr
 - [x] OptionalTryExpr
 - [x] PrefixUnaryExpr
 - [x] PostfixUnaryExpr
 - [ ] TupleElementExpr
 - [ ] InterpolatedStringLiteralExpr
 - [ ] TapExpr
 - [ ] SuperRefExpr
 - [ ] DynamicTypeExpr
 - [ ] ObjCSelectorExpr
 - [ ] ObjectLiteralExpr
 - [ ] MagicIdentifierLiteralExpr
 - [ ] DynamicMemberRefExpr
 - [ ] DynamicSubscriptExpr
 - [ ] EnumIsCaseExpr
 - [ ] KeyPathDotExpr
 - [ ] UnresolvedMemberExpr
 - [ ] EditorPlaceholderExpr
 - [ ] OverloadedDeclRefExpr
 - [ ] SequenceExpr
 - [x] UnresolvedDeclRefExpr
 - [ ] UnresolvedPatternExpr
 - [ ] DotSyntaxBaseIgnoredExpr
 - [ ] OtherConstructorDeclRefExpr
 - [ ] RebindSelfInConstructorExpr
 - [ ] MakeTemporarilyEscapableExpr
 - [ ] All the ImplicitConversionExpr subclasses

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5350,23 +5538,37 @@ void PrintAST::visitDestructureTupleExpr(DestructureTupleExpr *expr) {
 }
 
 void PrintAST::visitEditorPlaceholderExpr(EditorPlaceholderExpr *expr) {
+  if (expr->isTypeReference()) {
+    printTypeLoc(expr->getPlaceholderTypeRepr());
+  } else {
+    Printer << "<#";
+    Printer << expr->getPlaceholder();
+    Printer << "#>";
+  }
```
I actually think you only need to print `getPlaceholder()` here (for both cases) since that should include the entire `<#...#>` string

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5350,23 +5538,37 @@ void PrintAST::visitDestructureTupleExpr(DestructureTupleExpr *expr) {
 }
 
 void PrintAST::visitUnresolvedMemberExpr(UnresolvedMemberExpr *expr) {
+  Printer << ".";
+  Printer << expr->getName();
```
Looks good! Could you add a test please? :)

Rajveer100 commented:
I am facing a small issue related to `LLDB`:

```
**An executable path doesn't exist.**

The local path is “(null)”.

LLDB reports that it doesn't exist. Check that the executable remains on disk.
```

Never faced this issue earlier.

Rajveer100 commented:
PS: I fixed it, I had actually `migrated` the project to a different `directory` so the executable had to be changed.

Rajveer100 commented:
Could you provide `tests` for these `expressions`, I believe I have added tests for rest all expressions:

- DynamicMemberRefExpr
- EditorPlaceholderExpr
- OverloadedDeclRefExpr
- DotSyntaxBaseIgnoredExpr

For KeyPathExpr, I had added the tests earlier, although let me know if I am missing out on any specific cases.

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5455,23 +5647,34 @@ void PrintAST::visitDestructureTupleExpr(DestructureTupleExpr *expr) {
 
 void PrintAST::visitDynamicSubscriptExpr(DynamicSubscriptExpr *expr) {
+  visit(expr->getBase());
+  printArgumentList(expr->getArgs()->getOriginalArgs(), true);
+  // forSubscript - is `false` by default
```
An inline comment would be better for these e.g
```suggestion
  printArgumentList(expr->getArgs()->getOriginalArgs(), /*forSubscript*/ true);
```

hamishknight commented on test/expr/print/misc_expr.swift:
```diff
@@ -0,0 +1,123 @@
+)
+
+
+let userRoleKeyPath = \User.role
+// CHECK: @_hasInitialValue internal let userRoleKeyPath: KeyPath<User, Role> = \User
```
This should be checking for `\User.role`, remember we need to be printing out the original code.

Some other key path tests you could add are e.g:


```swift
// Multiple members
\X.y.z

// Optional chaining
\X.y?.z

// Optional force
\X.y!
\X.y!.z

// Dot self
\X.self

// Subscript
\X[0]
\X.[0]
\X.y[0]

// Tuple
\(Int, Int).0
\(x: Int, y: Int).x
```

hamishknight commented:
> Could you provide tests for these expressions, I believe I have added tests for rest all expressions:
> 
> - DynamicMemberRefExpr

This is for AnyObject lookup, something like e.g:

```swift

class C {
  @objc
  var foo = 0
}

func foo(_ x: AnyObject) {
  let y = x.foo
}
```

> - EditorPlaceholderExpr

I gave you examples of this in https://github.com/apple/swift/pull/64667#issuecomment-1674524262 :)

> This is for an editor placeholder `<#placeholder#>`, they can also include a type, in which case they print as `<#T##placeholder##Type#>` where placeholder can be any text, and Type is a type.

> - OverloadedDeclRefExpr

This is only present in un-typechecked code, so I wouldn't worry about it (I don't think we currently allow printing AST for invalid code)

> - DotSyntaxBaseIgnoredExpr

This is used for e.g curried references, e.g:

```swift
struct S {
  func foo() {}
}
let fn = S.foo
```

Rajveer100 commented:
> This is used for e.g curried references, e.g:

For `DotSyntaxBaseIgnoredExpr`, how do we avoid printing `self`, at the moment I print the `LHS` and `RHS` and I get this:

```Swift
S.self.foo()
```

Rajveer100 commented:
@hamishknight 
Let's wrap this!

hamishknight commented:
Sorry for dropping the ball on this, I'll try have a look this week

slavapestov commented:
@swift-ci Please smoke test

Rajveer100 commented:
> Given that this PR has been open for so long and nothing depends on ASTPrinter output of expressions, we should probably merge it. No reason it has to be perfect before merging. This would be useful to have when debugging and I'd rather have the option of imperfect output than just the s-expression ASTDumper output! :)

Sounds good, the remaining work can be continued as you said!

Rajveer100 commented:
I think apart from `TapExpr` and it's dependent `visit functions`, we could finish out the rest?

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5310,7 +5371,79 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+    Printer << ")";
+  } else if (auto rootType = expr->getRootType()) {
+    Printer << "\\";
+    printTypeLoc(rootType);
```
```suggestion
    printType(rootType);
    // FIXME: Print components
```

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -5335,61 +5475,110 @@ void PrintAST::visitMemberRefExpr(MemberRefExpr *expr) {
 }
 
 void PrintAST::visitCaptureListExpr(CaptureListExpr *expr) {
+  auto captureList = expr->getCaptureList();
```
```suggestion
  // FIXME: This should be moved into `printClosure` and merged with the closure
  // printing implementation.
  auto captureList = expr->getCaptureList();
```

hamishknight commented on test/expr/print/conditions.swift:
```diff
@@ -36,3 +36,23 @@ repeat {
+}
+
+var tryExpr = try? someThrowingFunc()
+// CHECK: @_hasInitialValue internal var tryExpr: someError? = try? someThrowingFunc()
```
```suggestion
// CHECK: @_hasInitialValue internal var tryExpr: SomeError? = try? someThrowingFunc()
```

hamishknight commented on test/expr/print/misc_expr.swift:
```diff
@@ -0,0 +1,140 @@
+// RUN: %target-swift-frontend -print-ast %s 2>&1 | %FileCheck %s
```
```suggestion
// RUN: %target-swift-frontend -print-ast -disable-objc-attr-requires-foundation-module %s 2>&1 | %FileCheck %s
```

hamishknight reviewed (commented):
Happy to land this as-is, although could you add a couple of FIXME comments for KeyPathExpr and CaptureListExpr since those are currently incomplete? I've also left a quick fix for EditorPlaceholderExpr.

hamishknight commented:
Note you will also need to fixup `test/expr/print/protocol.swift` to drop the access modifiers from the local vars:

```diff
  // CHECK: internal func cast(_ a: Any) {
- // CHECK:   @_hasInitialValue private let conditional: (any Archivable)? = a as? any Archivable
- // CHECK:   @_hasInitialValue private let forced: any Archivable = a as! any Archivable
+ // CHECK:   @_hasInitialValue let conditional: (any Archivable)? = a as? any Archivable
+ // CHECK:   @_hasInitialValue let forced: any Archivable = a as! any Archivable
  // CHECK: }
```

Rajveer100 commented:
`KeyPathExpr` is complete right, or there some significant changes to do still?

hamishknight commented:
> `KeyPathExpr` is complete right, or are there some significant changes to do still?

See https://github.com/apple/swift/pull/64667#discussion_r1384793411, the components still need printing

Rajveer100 commented:
> > `KeyPathExpr` is complete right, or are there some significant changes to do still?
> 
> See [#64667 (comment)](https://github.com/apple/swift/pull/64667#discussion_r1384793411), the components still need printing

Under `printKeyPathComponents`, which components are missing?

hamishknight commented:
> Under `printKeyPathComponents`, which components are missing?

The individual components are good, but you're not calling into `printKeyPathComponents` for regular key paths, and you're missing the dots between components.

Rajveer100 commented:
> The individual components are good, but you're not calling into `printKeyPathComponents` for regular key paths, and you're missing the dots between components.

IIUC, we need to call this irrespective of any of the conditions (like ObjC or if root exists or not)?

hamishknight commented:
> IIUC, we need to call this irrespective of any of the conditions (like ObjC or if root exists or not)?

Yes, that's correct. FWIW it looks like the `getParsedRoot()` path is also missing the leading `\`. I would do the `expr->isObjC()` check as an early return, then print the `\`, then have the if statement where you print the root, then print the components. Happy for this to be done in a follow-up though if you leave it as a FIXME.

hamishknight commented:
@swift-ci please test

Rajveer100 commented:
Tests pass locally, I have pushed again, maybe a `rebase` should solve this, as there's few other tests not part of my changes that also fail?

hamishknight commented on test/expr/print/misc_expr.swift:
```diff
@@ -0,0 +1,140 @@
+// RUN: %target-swift-frontend -print-ast -disable-objc-attr-requires-foundation-module %s 2>&1 | %FileCheck %s
```
```suggestion
// RUN: %target-swift-frontend -print-ast -disable-objc-attr-requires-foundation-module -enable-objc-interop %s 2>&1 | %FileCheck %s
```

hamishknight reviewed (commented):
Looks like you need to make a couple of other changes (the change to not print access modifiers for local declarations has affected a couple of other tests):

In `test/attr/accessibility_print.swift`:

```diff
@@ -211,14 +211,14 @@ public extension FE_PublicClass {
 
 // CHECK-LABEL: internal func GA_localTypes()
 func GA_localTypes() {
-  // CHECK-SRC: private struct Local {
+  // CHECK-SRC: struct Local {
   struct Local {
     // CHECK-SRC: internal let x
     let x = 0
   }
   _ = Local()
 
-  // CHECK-SRC: private enum LocalEnum {
+  // CHECK-SRC: enum LocalEnum {
   enum LocalEnum {
     // CHECK-SRC: {{^}} case A
     case A, B
```

In `decl/enum/derived_hashable_equatable.swift`:

```diff
@@ -8,14 +8,14 @@ enum Simple: Hashable {
   case b
 
   // CHECK:        @_implements(Equatable, ==(_:_:)) internal static func __derived_enum_equals(_ a: Simple, _ b: Simple) -> Bool {
-  // CHECK-NEXT:     private var index_a: Int
+  // CHECK-NEXT:     var index_a: Int
   // CHECK-NEXT:     switch a {
   // CHECK-NEXT:     case .a:
   // CHECK-NEXT:       index_a = 0
   // CHECK-NEXT:     case .b:
   // CHECK-NEXT:       index_a = 1
   // CHECK-NEXT:     }
-  // CHECK-NEXT:     private var index_b: Int
+  // CHECK-NEXT:     var index_b: Int
   // CHECK-NEXT:     switch b {
   // CHECK-NEXT:     case .a:
   // CHECK-NEXT:       index_b = 0
@@ -26,7 +26,7 @@ enum Simple: Hashable {
   // CHECK-NEXT:   }
 
   // CHECK:        internal func hash(into hasher: inout Hasher) {
-  // CHECK-NEXT:     private var discriminator: Int
+  // CHECK-NEXT:     var discriminator: Int
   // CHECK-NEXT:     switch self {
   // CHECK-NEXT:     case .a:
   // CHECK-NEXT:       discriminator = 0
@@ -101,14 +101,14 @@ enum HasUnavailableElement: Hashable {
   case b
 
   // CHECK:       @_implements(Equatable, ==(_:_:)) internal static func __derived_enum_equals(_ a: HasUnavailableElement, _ b: HasUnavailableElement) -> Bool {
-  // CHECK-NEXT:    private var index_a: Int
+  // CHECK-NEXT:    var index_a: Int
   // CHECK-NEXT:    switch a {
   // CHECK-NEXT:    case .a:
   // CHECK-NEXT:      index_a = 0
   // CHECK-NEXT:    case .b:
   // CHECK-NEXT:      _diagnoseUnavailableCodeReached{{.*}}()
   // CHECK-NEXT:    }
-  // CHECK-NEXT:    private var index_b: Int
+  // CHECK-NEXT:    var index_b: Int
   // CHECK-NEXT:    switch b {
   // CHECK-NEXT:    case .a:
   // CHECK-NEXT:      index_b = 0
@@ -119,7 +119,7 @@ enum HasUnavailableElement: Hashable {
   // CHECK-NEXT:  }
 
   // CHECK:       internal func hash(into hasher: inout Hasher) {
-  // CHECK-NEXT:    private var discriminator: Int
+  // CHECK-NEXT:    var discriminator: Int
   // CHECK-NEXT:    switch self {
   // CHECK-NEXT:    case .a:
   // CHECK-NEXT:      discriminator = 0
@@ -185,14 +185,14 @@ enum UnavailableEnum: Hashable {
   case b
 
   // CHECK:        @_implements(Equatable, ==(_:_:)) internal static func __derived_enum_equals(_ a: UnavailableEnum, _ b: UnavailableEnum) -> Bool {
-  // CHECK-NEXT:     private var index_a: Int
+  // CHECK-NEXT:     var index_a: Int
   // CHECK-NEXT:     switch a {
   // CHECK-NEXT:     case .a:
   // CHECK-NEXT:       index_a = 0
   // CHECK-NEXT:     case .b:
   // CHECK-NEXT:       index_a = 1
   // CHECK-NEXT:     }
-  // CHECK-NEXT:     private var index_b: Int
+  // CHECK-NEXT:     var index_b: Int
   // CHECK-NEXT:     switch b {
   // CHECK-NEXT:     case .a:
   // CHECK-NEXT:       index_b = 0
@@ -203,7 +203,7 @@ enum UnavailableEnum: Hashable {
   // CHECK-NEXT:   }
 
   // CHECK:        internal func hash(into hasher: inout Hasher) {
-  // CHECK-NEXT:     private var discriminator: Int
+  // CHECK-NEXT:     var discriminator: Int
   // CHECK-NEXT:     switch self {
   // CHECK-NEXT:     case .a:
   // CHECK-NEXT:       discriminator = 0
```

hamishknight commented:
@swift-ci please test

Rajveer100 commented:
Regressions macOS Platform:

```
Failed Tests (2):
  Swift(macosx-x86_64) :: decl/enum/derived_hashable_equatable_macos.swift
  Swift(macosx-x86_64) :: expr/print/func_closures.swift
```

hamishknight commented:
@swift-ci please test macOS

hamishknight commented:
Looks like a couple more test tweaks are needed:

`test/decl/enum/derived_hashable_equatable_macos.swift`:

```diff
@@ -25,7 +25,7 @@ enum HasElementsWithAvailability: Hashable {
   case introduced10_50
 
   // CHECK:       @_implements(Equatable, ==(_:_:)) internal static func __derived_enum_equals(_ a: HasElementsWithAvailability, _ b: HasElementsWithAvailability) -> Bool {
-  // CHECK-NEXT:    private var index_a: Int
+  // CHECK-NEXT:    var index_a: Int
   // CHECK-NEXT:    switch a {
   // CHECK-NEXT:    case .alwaysAvailable:
   // CHECK-NEXT:      index_a = 0
@@ -40,7 +40,7 @@ enum HasElementsWithAvailability: Hashable {
   // CHECK-NEXT:    case .introduced10_50:
   // CHECK-NEXT:      index_a = 2
   // CHECK-NEXT:    }
-  // CHECK-NEXT:    private var index_b: Int
+  // CHECK-NEXT:    var index_b: Int
   // CHECK-NEXT:    switch b {
   // CHECK-NEXT:    case .alwaysAvailable:
   // CHECK-NEXT:      index_b = 0
@@ -59,7 +59,7 @@ enum HasElementsWithAvailability: Hashable {
   // CHECK-NEXT:  }
 
   // CHECK:       internal func hash(into hasher: inout Hasher) {
-  // CHECK-NEXT:    private var discriminator: Int
+  // CHECK-NEXT:    var discriminator: Int
   // CHECK-NEXT:    switch self {
   // CHECK-NEXT:    case .alwaysAvailable:
   // CHECK-NEXT:      discriminator = 0
```

`test/expr/print/func_closures.swift`:

```diff
@@ -1,4 +1,4 @@
-// RUN: %target-swift-frontend -print-ast %s 2>&1 | %FileCheck %s
+// RUN: %target-swift-frontend -print-ast -disable-availability-checking %s 2>&1 | %FileCheck %s
 
 func fetch() async throws -> String {
 }
```

hamishknight commented:
@swift-ci please test

Rajveer100 commented:
CI looks good!

hamishknight reviewed (approved):
Thank you!

Rajveer100 commented:
Welcome!
